CorsixTH/Lua/objects/reception_desk.lua

```diff
@@ -92,48 +92,74 @@ function ReceptionDesk:onClick(ui, button)
+    if inspector.going_home then return end
+    local epidemic = self.hospital.epidemic
+    if epidemic then
+      -- The result of the epidemic may already by determined
+      -- i.e if an infected patient has left the hospital
+      if not epidemic.result_determined then
+        epidemic:finishCoverUp()
+      end
+      epidemic:applyOutcome()
+      inspector:goHome()
+    end
+  end
+
   local queue_front = self.queue:front()
-  local reset_timer = true
-  if self.receptionist and queue_front then
+  --! Advance the front humanoid of the queue.
+  --! Returns true if ready to be dealt with.
+  local function advanceQueue()
     if queue_front:getCurrentAction().name == "idle" then
       self.queue_advance_timer = self.queue_advance_timer + 1
+    end
+    return self.queue_advance_timer >= 4 + Date.hoursPerDay() * (1.0 - self.receptionist.profile.skill)
+  end
+
+  local reset_timer = true
+  if self.receptionist and queue_front then
+    if not advanceQueue() then
+      -- Not ready to deal with patient, continue waiting
       reset_timer = false
-      if self.queue_advance_timer >= 4 + Date.hoursPerDay() * (1.0 - self.receptionist.profile.skill) then
-        reset_timer = true
-        if queue_front.next_room_to_visit then
-          queue_front:queueAction(SeekRoomAction(queue_front.next_room_to_visit.room_info.id))
-        else
-          if class.is(queue_front, Inspector) then
-            local inspector = queue_front
-            if not inspector.going_home  then
-              local epidemic = self.hospital.epidemic
-              if epidemic then
-                -- The result of the epidemic may already by determined
-                -- i.e if an infected patient has left the hospital
-                if not epidemic.result_determined then
-                  epidemic:finishCoverUp()
-                end
-                epidemic:applyOutcome()
-                inspector:goHome()
-              end
-            end
-            -- VIP has his own list, don't add the gp office twice
-          elseif queue_front.humanoid_class ~= "VIP" then
-            if queue_front:agreesToPay("diag_gp") then
-              queue_front:queueAction(SeekRoomAction("gp"))
-            else
-              queue_front:goHome("over_priced", "diag_gp")
-            end
-          else
-            -- the VIP will realise that he is idle, and start going round rooms
-            queue_front:queueAction(IdleAction())
-            queue_front.waiting = 1
-          end
-        end
-        self.queue:pop()
-        self.queue.visitor_count = self.queue.visitor_count + 1
-        queue_front.has_passed_reception = true
-      end
+    elseif queue_front.next_room_to_visit then
+      handlePatient(queue_front, false)
+    elseif class.is(queue_front, Inspector) then
+      handleInspector(queue_front)
+    elseif class.is(queue_front, Vip) then
+      handleVIP(queue_front)
+    else -- New patient
+      handlePatient(queue_front, true)
```

Comment:
I've passed `queue_front` here into the respective functions, rather than declaring `queue_front` at the top, to make those functions more fitting to context.

Reply:
Very good!